content/sensu-core/0.29/guides/pre-compile-plugins.md

```diff
@@ -0,0 +1,140 @@
+---
+title: "Pre-Compiling Plugins"
+description: "Pre-Compiling Sensu Plugins for Distribution"
+product: "Sensu Core"
+version: "0.29"
+weight: 2
+next: ../intro-to-checks
+previous: ../overview
+menu:
+  sensu-core-0.29:
+    parent: guides
+---
+
+# Overview
+
+This guide will walk you through an example on how to pre-compile Sensu Plugins for distribution among clients.
+The goal of this guide is to showcase how you can quickly install plugins on ephemeral instances where time to compile can affect time to production  .
+We'll be using `sensu-plugins-aws` as it requires `nokogiri`,  usually requiring a longer build time.
+
+# Prerequist
+
+Sources:
+
+* [GitHub - luislavena/gem-compiler: A RubyGems plugin that generates binary gems](https://github.com/luislavena/gem-compiler)
+
+This guide assumes the following:
+
+* Centos 7 platform
+* You have some way to store these artifacts and make them available for systems in your environment
+
+_NOTE: Compiled gems are tied to the Ruby version they are compiled for. If Sensu upgrade Ruby 2.4 -> Ruby 2.5, packages will need to be rebuilt._
+_NOTE: Compiled gems are tied to CPU architecture for which they are compiled on._
+
+## Installing The Build Toolchain
+
+We'll first add the Sensu yum repository definition and install Sensu
+
+{{< highlight shell >}}
+printf '[sensu]
+name=sensu
+baseurl=https://repositories.sensuapp.org/yum/$releasever/$basearch/
+gpgcheck=0
+enabled=1' | sudo tee /etc/yum.repos.d/sensu.repo && sudo yum install -y sensu{{< /highlight >}}
+
+And then install EPEL, Dev tools and libs required for the next steps
+
+{{< highlight shell >}}
+sudo yum install -y epel-release libxml2 libxslt
+sudo yum groupinstall -y "Development Tools"{{< /highlight >}}
+
+Finally we'll add Sensu's embedded bin to our path
+
+{{< highlight shell >}}
+export PATH=$PATH:/opt/sensu/embedded/bin{{< /highlight >}}
+
+## Installing a Gem to a tmp directory
+
+Install the gem-compiler package
+
+{{< highlight shell >}}
+gem install gem-compiler{{< /highlight >}}
+
+We'll now install the source gems into a cache directory
+
+{{< highlight shell >}}
+mkdir /tmp/gems
+gem install --no-ri --no-rdoc --install-dir /tmp/gems sensu-plugins-aws
+gem install --no-ri --no-rdoc --install-dir /tmp/gems mini_portile2{{< /highlight >}}
+
+Observing the `gem install` output, we can see that nokogiri and unf_ext build native extensions.
+These are the gems we will compile so we can install them from our own repository.
+
+### Compile the Gems for packaging
+
+We'll use `gem-compiler` to rebuild the gem packages with the included native extensions:
+
+{{< highlight shell >}}
+cd /tmp/gems/cache
+gem compile unf_ext-*.gem
+gem compile nokogiri-1.8.*.gem --prune{{< /highlight >}}
```

Comment:
I get this error:

```
Unpacking gem: 'nokogiri-1.8.4' in temporary directory...
Building native extensions.  This could take a while...
ERROR:  While executing gem ... (Gem::Ext::BuildError)
    ERROR: Failed to build gem native extension.

    current directory: /tmp/d20180926-13682-1ip9vmd/nokogiri-1.8.4/ext/nokogiri
/opt/sensu/embedded/bin/ruby -r ./siteconf20180926-13682-61ubeq.rb extconf.rb
checking if the C compiler accepts ... yes
Building nokogiri using packaged libraries.
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/opt/sensu/embedded/bin/$(RUBY_BASE_NAME)
	--help
	--clean
	--use-system-libraries
/opt/sensu/embedded/lib/ruby/site_ruby/2.4.0/rubygems/dependency.rb:308:in `to_specs': Could not find 'mini_portile2' (~> 2.3.0) among 52 total gem(s) (Gem::MissingSpecError)
Checked in 'GEM_PATH=/home/vagrant/.gem/ruby/2.4.0:/opt/sensu/embedded/lib/ruby/gems/2.4.0', execute `gem env` for more information
	from /opt/sensu/embedded/lib/ruby/site_ruby/2.4.0/rubygems/dependency.rb:320:in `to_spec'
	from /opt/sensu/embedded/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_gem.rb:65:in `gem'
	from extconf.rb:463:in `<main>'

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /tmp/d20180926-13682-1ip9vmd/nokogiri-1.8.4/lib/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /tmp/d20180926-13682-1ip9vmd/nokogiri-1.8.4 for inspection.
Results logged to /tmp/d20180926-13682-1ip9vmd/nokogiri-1.8.4/lib/gem_make.out
```